### PR TITLE
Further Structure API Enhancements

### PR DESCRIPTION
**Pull request by APickledWalrus**
### Description
This PR attempts to further improve some of the new API from the Structure API PR.

The point on ScriptData was not fully addressed. Unfortunately, I do not think there is a good way for us to automatically clear ScriptData during unload. We could remove all ScriptData after the unloading process, but the Script will be removed at that point and GCed (so I don't think it really makes a difference). I think our best option may be to continue on the assumption that users of ScriptData will clean up their data before unloading is fully complete (if necessary). Areas depending on ScriptData provided by a Structure or other element should continue to perform null checks to ensure that the data is actually present.

---
**Target Minecraft Versions:** Any
**Requirements:** N/A
**Related Issues:**
- #5318 
- #5372 (fixes an issue with the varargs method update from that PR)

### CODE REVIEW

File: src/main/java/ch/njol/skript/structures/StructOptions.java [hunk 1]
```diff
@@ -113,44 +113,33 @@ public String toString(@Nullable Event e, boolean debug) {
...
-		});
-
-		assert replaced != null;
-		return replaced;
-	}
-
-	private static final class OptionsData implements ScriptData {
+	public static final class OptionsData implements ScriptData {
+
+		private final HashMap<String, String> options = new HashMap<>(15);
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/structures/StructOptions.java:118]:
```suggestion
		private final Map<String, String> options = new HashMap<>(15);
```

**Review by TheLimeGlass — APPROVED** [review]

File: src/main/java/ch/njol/skript/structures/StructOptions.java (hunk 1 above)
Comment (kiip1) [reply src/main/java/ch/njol/skript/structures/StructOptions.java:118]: Nit-pick but entering 15 as parameter is redundant as the default is 16.

File: src/main/java/ch/njol/skript/ScriptLoader.java [hunk 2]
```diff
@@ -798,12 +791,17 @@ private static Config loadStructure(InputStream source, String name) {
 	 * @return Combined statistics for the unloaded scripts.
 	 *         This data is calculated by using {@link ScriptInfo#add(ScriptInfo)}.
 	 */
-	public static ScriptInfo unloadScripts(Set<Script> scripts) {
+	public static ScriptInfo unloadScripts(Collection<Script> scripts) {
```
[the remark this answers is not in the document]
Comment (kiip1) [reply src/main/java/ch/njol/skript/ScriptLoader.java:794]: Could also do:
```suggestion
	public static ScriptInfo unloadScripts(Collection<Script> scripts) {
		scripts = new HashSet<>(scripts);
```

File: src/main/java/ch/njol/skript/ScriptLoader.java [hunk 3]
```diff
@@ -493,18 +493,17 @@ private static CompletableFuture<ScriptInfo> loadScripts(List<Config> configs, O
 		
 		ScriptInfo scriptInfo = new ScriptInfo();
 
-		List<Script> scripts = new ArrayList<>();
+		List<NonNullPair<Script, List<Structure>>> scripts = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/ScriptLoader.java:496]: Agree, could be Multimap.

File: src/main/java/org/skriptlang/skript/lang/script/Script.java [hunk 4]
```diff
@@ -41,19 +40,20 @@ public final class Script {
...
 	 * Creates a new Script to be used across the API.
-	 * Only one script should be created per Config. A loaded script may be obtained through {@link ch.njol.skript.ScriptLoader}.
-	 * @param config The Config containing the contents of this script.
+	 * Only one Script should be created per Config. A loaded Script may be obtained through {@link ch.njol.skript.ScriptLoader}.
+	 * @param config The Config containing the contents of this Script.
 	 */
-	public Script(Config config) {
+	public Script(Config config, List<Structure> structures) {
 		this.config = config;
+		this.structures = structures;
```
[the remark this answers is not in the document]
Comment (TheLimeGlass) [reply src/main/java/org/skriptlang/skript/lang/script/Script.java:52]: could also set the field up originally
```java
private final List<Structure> structures = new ArrayList<>();

//init... {
    this.structures.addAll(structures);
}

File: src/main/java/ch/njol/skript/ScriptLoader.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (APickledWalrus) [reply src/main/java/ch/njol/skript/ScriptLoader.java:496]: Don't we lose the reference to the List of Structures with a Multimap? This is kind of yucky, but we _need_ a reference to the modifiable version of the Script's Structure list.

[the remark this answers is not in the document]
Comment (APickledWalrus) [reply src/main/java/ch/njol/skript/ScriptLoader.java:495]: Just to note here: We had discussed on Discord about what we might be able to do with this. Some alternatives were discussed, but our best options likely won't happen without some recodes (and taking advantage of features like package private methods). I think we are okay to ignore this for now as it's purely internal API and can easily be changed in the future without causing a major disruption. Thus, I'm going to mark this as resolved - please feel free to unresolve if your position has changed/you have a new idea.

**Review by kiip1 — APPROVED** [review]

**Review by AyhamAl-Ali — APPROVED** [review]
Looking good 🚀

**Pull request merged**